helper/main.cpp

```diff
@@ -17,31 +17,55 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-#include <QCoreApplication>
+#include <stdexcept>
+
 #include <QString>
 #include <QTextStream>
 #include <QTranslator>
 
-#include "restorejob.h"
-#include "writejob.h"
+// Platform specific drive handler.
+#include "drive.h"
+#include "write.h"
+
+void restore(Drive *const drive) {
+    drive->umount();
+    drive->wipe();
+    drive->addPartition();
+}
 
 int main(int argc, char *argv[]) {
-    QCoreApplication app(argc, argv);
+    const QString action = argv[1];
+    bool isRestore = argc == 3 && action == "restore";
+    bool isWrite = argc == 4 && action == "write";
+    QTextStream err(stderr);
+    if (!isRestore && !isWrite) {
+        err << "Helper: Wrong arguments entered\n";
+        err.flush();
+        return 1;
+    }
 
     QTranslator translator;
     translator.load(QLocale(), QString(), QString(), ":/translations");
-    app.installTranslator(&translator);
 
-    if (app.arguments().count() == 3 && app.arguments()[1] == "restore") {
-        new RestoreJob(app.arguments()[2]);
-    }
-    else if (app.arguments().count() == 4 && app.arguments()[1] == "write") {
-        new WriteJob(app.arguments()[2], app.arguments()[3]);
-    }
-    else {
-        QTextStream err(stderr);
-        err << "Helper: Wrong arguments entered";
+    QString driveIdentifier = isRestore ? argv[2] : argv[3];
```

Comment:
I guess the manual cancel could have been avoided in an other way. It's not, uh, Qt to use exceptions. Anyway, I think we can just change this any time in the future.

Also, please don't take performance cost in such scenarios into account at all. Readability of the code trumps all if it's not a performance-critical piece of code.

In case this was a repeated action, performance would matter - for example in the part that computes the checksum. These exceptions are however something that happens at most once in the lifetime of the application which makes the benefit so negligible we can't even care about it.